### PR TITLE
feat: overhaul encumbrance UI

### LINKED ISSUE #3422: ☂️ Armor coverage UI issues
### Describe the bug

This. This thing.
image

In a nutshell:
1. The numbers aren't aligned properly *ewwww*
2. Average coverage is kinda really gonna be useless 90% of the time. Even if coverage actually varies, a high average means absolutely nothing to the player if it's because the limbs have perfect coverage but the vitals are highly exposed, or vice-versa.
3. Large volume of redundant information anytime we have an item that doesn't actually vary its encumbrance and coverage percent 

What needs to be fixed for this, in order:
1. Fix the spacing on the display info weh!
2. Either axe average coverage because it's going to be not really that useful to the player most of the time, or else only display it if the actual average is different from all of the values for each body part (though as noted above that's both the only case where it's worth displaying but also the case where it's the most worthless to the player).
3. If all bodyparts have the exact same info, don't display a detail view for every single part. Just show a "everything has this" line instead.
 3a. Addition to the above, in theory we could even make the detail view even better by grouping all examples found of identical values, e.g. if the torso and head has 100% coverage while the arms and hands have 90% coverage, find some (hopefully not ugly) way to say "head, torso: 100" followed by "arms, hands: 90"

### Steps To Reproduce

Exist in a build after and look at the description screen for any armor item, mainly anything that covers more than one body part.

### Screenshots

_No response_

### Versions and configuration

- OS: Windows
    - OS Version: 10.0.19041.450 (2004)
- Game Version: 2bc98a8 [64-bit]
- Graphics Version: Tiles
- LAPI Version: 1
- Game Language:  []
- Mods loaded: [
    Bright Nights [bn],
    Arcana and Magic Items [Arcana],
    Cataclysm++ [Cata++],
    Elevated bridges [elevated_bridges],
    Medieval Mod Reborn [Medieval_Mod_Reborn],
    MST Extra [MST_Extra],
    Arcana/Cata++ Patchmod [Arcana_Cata++_Patch],
    Mining Mod [Mining_Mod],
    Tankmod: Revived [Tankmod_Revived],
    No Fungal Monsters [No_Fungi],
    Disable NPC Needs [no_npc_food],
    Faster Stamina Regen [stamina_regen_buff],
    Manual Bionic Installation [manualbionicinstall],
    Nonperishable Overhaul [Nonperishable_Overhaul],
    Simplified Nutrition [novitamins],
    SpeedyDex [speedydex],
    Stats Through Kills [stats_through_kills]
]

### Additional context

DDA had an issue for exactly this problem way back when, with thanks to @x-qq: https://github.com/CleverRaven/Cataclysm-DDA/issues/59820

The UI shown in the issue OP is somehow even shittier than the version we've ended up with but still, ick. Some of it it seems is due to breathability which mercifully we don't have, but in any case it predictably got closed as invalid because since then they've managed to make their armor UI even more inefficient and hideous: 
image
image

Jesus fuckmothering christ I have to pagedown through that and take two separate screenshots to get all that and I've got DDA maximized on a 1920x1080 monitor what in the infinite hells
image

```[tasklist]
### Tasks
- [ ] https://github.com/cataclysmbnteam/Cataclysm-BN/issues/3436
- [ ] https://github.com/cataclysmbnteam/Cataclysm-BN/issues/3435
```

### LINKED ISSUE #3435: redundant info in armor coverage UI

### LINKED ISSUE #3436: misalignments in armor coverage UI

### PR DESCRIPTION
**Pull request by scarf005**
## Summary

SUMMARY: Interface "Overhauled encumbrance UI"

## Purpose of change

- fix #3436 
- fix #3422 
- fix #3435 
- release 0.4

## Describe the solution

1. extracted item armor display function to new files because `item.cpp` takes 30 seconds to rebuild each time.
2. calculated widest bodypart name to prevent any weird misalignments.
3. used utf8 padding to misalignment.
4. grouped body parts with same encumbrance and encumbrance.
5. simplified display when all body parts have same value.
6. removed average encumbrance display as it had little uses.

## Describe alternatives you've considered

~~using proper grouping and stacked bar display, but this already took too much time.~~ _hecc._

## Testing

Cataclysm: Bright Nights - 1d5705aa1b9a_01
Cataclysm: Bright Nights - 1d5705aa1b9a_02
Cataclysm: Bright Nights - 1d5705aa1b9a_03

### CODE REVIEW

**Comment by olanti-p** [comment]
It doesn't look like it's fully fixing #3422, just the alignment issue

**Comment by scarf005** [comment]
now it's completely reworked.

File: tests/iteminfo_test.cpp [hunk 1]
```diff
@@ -618,18 +573,18 @@ TEST_CASE( "armor coverage and protection values", "[item][iteminfo][armor]" )
...
-            "Average Coverage: <color_c_yellow>90</color>%  Warmth: <color_c_yellow>5</color>\n"
-            "--\n"
-            "<color_c_white>Encumbrance</color>: <color_c_red>(poor fit)</color>\n"
-            "Arms:  <color_c_yellow>3</color>  Coverage:  <color_c_yellow>90</color>\n"
-            "Torso:  <color_c_yellow>3</color>  Coverage:  <color_c_yellow>90</color>\n"
-            "<color_c_white>Protection</color>: Bash: <color_c_yellow>1</color>  Cut: <color_c_yellow>1</color>  Ballistic: <color_c_yellow>1</color>\n"
-            "  Acid: <color_c_yellow>0</color>  Fire: <color_c_yellow>0</color>  Environmental: <color_c_yellow>0</color>\n" );
+              R"(--
+<color_c_white>Covers</color>: The <color_c_cyan>torso</color>. The <color_c_cyan>arms</color>. )"
+            R"(
```
Comment (olanti-p) [inline tests/iteminfo_test.cpp:578]: Could you also fix the warnings tidy reports here?

If spaces are intentional, the lint can be silenced by adding `// NOLINTNEXTLINE(cata-text-style)` on the line before error line.

File: src/iteminfo_armor.cpp [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+    return result;
+}
+
+
+auto translate_pair( const BodyPartInfoPair &piece ) -> std::string
+{
+    return piece.second.to_display.translated();
+};
+
+using InfoGroup = std::map<int, std::vector<BodyPartInfoPair>>;
```
Comment (olanti-p) [inline src/iteminfo_armor.cpp:107]: Not used anywhere.

File: src/iteminfo_armor.cpp [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+const auto space = std::string {"  "};
+
+auto total()
+{
+    return  _( "<info>(total)</info>" );
+}
+
+auto item_coverages( const std::vector<BodyPartInfoPair> &xs )-> std::vector<iteminfo>
+{
+    const auto &grouped = cata::group_by( xs, []( const auto & info ) -> int {return info.second.portion.coverage;} );
```
Comment (olanti-p) [inline src/iteminfo_armor.cpp:128]: Astyle doesn't format this line for some reason. It feels like there should've been spaces before & after `return info.second.portion.coverage;` with our current style, but astyle neither adds nor remove them here if added.

File: src/iteminfo_armor.cpp [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+            if( sizing_info_str ) {
+                format = sizing_info_str.value();
+            }
+        }
+
+        if( const islot_armor *t = find_armor_data() ) {
+            if( !t->data.empty() ) {
+
+                std::map<bodypart_str_id, body_part_display_info> to_display_data;
+                const auto &avatar = get_avatar();
```
Comment (olanti-p) [inline src/iteminfo_armor.cpp:342]: 1. Avoid naming variables same as existing classes, it makes some compilers behave weirdly
2. There's already a reference to the avatar named `you` at the start of the function

File: src/iteminfo_armor.cpp [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+        if( sizing_matters ) {
+            const auto sizing_level = get_sizing( you );
+            const auto sizing_info_str = sizing_info( sizing_level );
+            if( sizing_info_str ) {
+                format = sizing_info_str.value();
+            }
+        }
+
+        if( const islot_armor *t = find_armor_data() ) {
+            if( !t->data.empty() ) {
```
Comment (olanti-p) [inline src/iteminfo_armor.cpp:339]: No need for extra if
```suggestion
        const islot_armor *t = find_armor_data();
        if( t && !t->data.empty() ) {
```

Upd: now that I'm looking through the code, the original PR added a lot of multi-level ifs that should've been `&&` instead...

Upd2: `t` is a horrible name in such context

File: src/iteminfo_armor.cpp [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+
+                std::map<bodypart_str_id, body_part_display_info> to_display_data;
+                const auto &avatar = get_avatar();
+                for( const armor_portion_data &piece : t->data ) {
+                    if( piece.covers.has_value() ) {
```
Comment (olanti-p) [inline src/iteminfo_armor.cpp:344]: Does `piece.covers` need to be a `std::optional<body_part_set>` instead of plain `body_part_set`?

**Review by olanti-p — COMMENTED** [review]
image
"Total" is misleading here. For example, for power armor it says total encumbrance is 50, while the actual total encumbrance is 50*5 = 250

**Comment by olanti-p** [comment]
Overall it's much better then before, no unnecessary verbosity for cases where it's not needed.

There's still a bit of inconsistency in spacing left:
image

On the topic of armor info, the protection values have bad formatting at times, but it's not related to this PR.
image

File: src/iteminfo_armor.cpp (hunk 6 above)
Comment (scarf005) [reply src/iteminfo_armor.cpp:344]: yeah, there's no reason for it to be when set has empty state.

**Comment by scarf005** [comment]
after applying requested changes

**Review by olanti-p — APPROVED** [review]

**Comment by scarf005** [comment]
yeah this looks better.

**Pull request merged**